File: 2020_solutions/cplusplus/main/day12.cpp
#include "day12.hpp"
#include <algorithm>
#include <stdlib.h>
//
// Created by luker on 2/2/2021.
//

namespace Day12 {
    enum class Action {
        NORTH,
        SOUTH,
        EAST,
        WEST,
        LEFT,
        RIGHT,
        FORWARD
    };
    enum class Direction {
        NORTH, SOUTH, EAST, WEST
    };

    class Ship {
        Direction direction = Direction::EAST;
        int32_t ship_x = 0;
        int32_t ship_y = 0;
        int32_t waypoint_x = 10;
        int32_t waypoint_y = 1;
    public:

        uint32_t distance_from_origin() const {
            return std::abs(ship_x) + std::abs(ship_y);
        }

        void add_ship_x(int32_t amount) {
            ship_x = ship_x + amount;
        }

        void add_ship_y(int32_t amount) {
            ship_y = ship_y + amount;
        }

        void add_waypoint_x(int32_t amount) {
            waypoint_x = waypoint_x + amount;
        }

        void add_waypoint_y(int32_t amount) {
            waypoint_y = waypoint_y + amount;
        }

        void add_ship_forward(int32_t amount) {
            switch (direction) {
                case Direction::EAST:
                    add_ship_x(amount);
                    return;
                case Direction::NORTH :
                    add_ship_y(amount);
                    return;
                case Direction::WEST :
                    add_ship_x(-1 * amount);
                    return;
                case Direction::SOUTH :
                    add_ship_y(-1 * amount);
                    return;
            }
        }

        void add_waypoint_forward(int32_t amount) {
            for (int32_t times = 0; times < amount; times++) {
                add_ship_x(waypoint_x);
                add_ship_y(waypoint_y);
            }

        }

        void ship_left() {
            switch (direction) {
                case Direction::EAST:
                    direction = Direction::NORTH;
                    return;
                case Direction::NORTH:
                    direction = Direction::WEST;
                    return;
                case Direction::WEST:
                    direction = Direction::SOUTH;
                    return;
                case Direction::SOUTH:
                    direction = Direction::EAST;
                    return;
            }
        }

        void ship_right() {
            switch (direction) {
                case Direction::EAST:
                    direction = Direction::SOUTH;
                    return;
                case Direction::NORTH:
                    direction = Direction::EAST;
                    return;
                case Direction::WEST:
                    direction = Direction::NORTH;
                    return;
                case Direction::SOUTH:
                    direction = Direction::WEST;
                    return;
            }
        }

        void waypoint_left() {
            int32_t new_x = -1 * waypoint_y;
            int32_t new_y = waypoint_x;

            waypoint_x = new_x;
            waypoint_y = new_y;
        }

        void waypoint_right() {
            int32_t new_x = waypoint_y;
            int32_t new_y = -1 * waypoint_x;

            waypoint_x = new_x;
            waypoint_y = new_y;
        }
    };

    class ShipInstruction {
        Action action;
        int32_t quantity;
    public:
        ShipInstruction(const std::string string) {
            switch (string[0]) {
                case 'N':
                    action = Action::NORTH;
                    break;
                case 'S':
                    action = Action::SOUTH;
                    break;
                case 'E':
                    action = Action::EAST;
                    break;
                case 'W':
                    action = Action::WEST;
                    break;
                case 'L':
                    action = Action::LEFT;
                    break;
                case 'R':
                    action = Action::RIGHT;
                    break;
                case 'F':
                    action = Action::FORWARD;
                    break;
                default:
                    action = Action::FORWARD;
            }
            quantity = static_cast<int32_t>(std::stol(string.substr(1)));
        }

        void apply_ship(Ship &ship) {
            const auto turns = quantity / 90;
            switch (action) {
                case Action::NORTH:
                    ship.add_ship_y(quantity);
                    break;
                case Action::SOUTH:
                    ship.add_ship_y(-1 * quantity);
                    break;
                case Action::EAST:
                    ship.add_ship_x(quantity);
                    break;
                case Action::WEST:
                    ship.add_ship_x(-1 * quantity);
                    break;
                case Action::FORWARD:
                    ship.add_ship_forward(quantity);
                    break;
                case Action::LEFT:
                    for (auto turn = 0; turn < turns; turn++) {
                        ship.ship_left();
                    }
                    break;
                case Action::RIGHT:
                    for (auto turn = 0; turn < turns; turn++) {
                        ship.ship_right();
                    }
                    break;
            }
        }

        void apply_waypoint(Ship &ship) {
            const auto turns = quantity / 90;
            switch (action) {
                case Action::NORTH:
                    ship.add_waypoint_y(quantity);
                    break;
                case Action::SOUTH:
                    ship.add_waypoint_y(-1 * quantity);
                    break;
                case Action::EAST:
                    ship.add_waypoint_x(quantity);
                    break;
                case Action::WEST:
                    ship.add_waypoint_x(-1 * quantity);
                    break;
                case Action::FORWARD:
                    ship.add_waypoint_forward(quantity);
                    break;
                case Action::LEFT:
                    for (auto turn = 0; turn < turns; turn++) {
                        ship.waypoint_left();
                    }
                    break;
                case Action::RIGHT:
                    for (auto turn = 0; turn < turns; turn++) {
                        ship.waypoint_right();
                    }
                    break;
            }
        }
    };

    uint32_t puzzle1(const std::vector<std::string> &inputValues) {
        Ship ship;

        std::for_each(inputValues.begin(), inputValues.end(), [&ship](std::string string) {
            ShipInstruction instruction(string);
            instruction.apply_ship(ship);
        });

        return ship.distance_from_origin();
    }

    uint32_t puzzle2(const std::vector<std::string> &inputValues) {
        Ship ship;

        std::for_each(inputValues.begin(), inputValues.end(), [&ship](std::string string) {
            ShipInstruction instruction(string);
            instruction.apply_waypoint(ship);
        });

        return ship.distance_from_origin();
    }
}

File: 2020_solutions/rust/src/day12.rs
enum Direction {
    NORTH,
    SOUTH,
    EAST,
    WEST,
}

enum Action {
    NORTH,
    SOUTH,
    EAST,
    WEST,
    LEFT,
    RIGHT,
    FORWARD,
}

struct Ship {
    direction: Direction,
    ship_x: i32,
    ship_y: i32,
    waypoint_x: i32,
    waypoint_y: i32,
}

impl Ship {
    fn new() -> Self {
        return Self {
            direction: Direction::EAST,
            ship_x: 0,
            ship_y: 0,
            waypoint_x: 10,
            waypoint_y: 1,
        };
    }

    fn distance_from_origin(&self) -> u32 {
        return (self.ship_x.abs() + self.ship_y.abs()) as u32;
    }

    fn add_ship_x(&mut self, amount: i32) {
        self.ship_x = self.ship_x + amount;
    }

    fn add_ship_y(&mut self, amount: i32) {
        self.ship_y = self.ship_y + amount;
    }

    fn add_waypoint_x(&mut self, amount: i32) {
        self.waypoint_x = self.waypoint_x + amount;
    }

    fn add_waypoint_y(&mut self, amount: i32) {
        self.waypoint_y = self.waypoint_y + amount;
    }

    fn add_ship_forward(&mut self, amount: i32) {
        match self.direction {
            Direction::EAST => self.add_ship_x(amount),
            Direction::NORTH => self.add_ship_y(amount),
            Direction::WEST => self.add_ship_x(-1 * amount),
            Direction::SOUTH => self.add_ship_y(-1 * amount),
        }
    }

    fn add_waypoint_forward(&mut self, amount: i32) {
        for times in 0..amount {
            self.add_ship_x(self.waypoint_x);
            self.add_ship_y(self.waypoint_y);
        }

    }

    fn ship_left(&mut self) {
        self.direction = match self.direction {
            Direction::EAST => Direction::NORTH,
            Direction::NORTH => Direction::WEST,
            Direction::WEST => Direction::SOUTH,
            Direction::SOUTH => Direction::EAST,
        }
    }

    fn ship_right(&mut self) {
        self.direction = match self.direction {
            Direction::EAST => Direction::SOUTH,
            Direction::NORTH => Direction::EAST,
            Direction::WEST => Direction::NORTH,
            Direction::SOUTH => Direction::WEST,
        }
    }

    fn waypoint_left(&mut self) {
        let new_x = -1 * self.waypoint_y;
        let new_y = self.waypoint_x;

        self.waypoint_x = new_x;
        self.waypoint_y = new_y;
    }

    fn waypoint_right(&mut self) {
        let new_x = self.waypoint_y;
        let new_y = -1 * self.waypoint_x;

        self.waypoint_x = new_x;
        self.waypoint_y = new_y;
    }
}

struct ShipInstruction {
    action: Action,
    quantity: i32,
}

impl ShipInstruction {
    fn new(string: &String) -> Self {
        let first_char = string.chars().next().unwrap();
        let quantity = &string[1..].parse::<i32>().unwrap();

        let action = match first_char {
            'N' => Action::NORTH,
            'S' => Action::SOUTH,
            'E' => Action::EAST,
            'W' => Action::WEST,
            'L' => Action::LEFT,
            'R' => Action::RIGHT,
            'F' => Action::FORWARD,
            _ => Action::FORWARD
        };

        return Self {
            action,
            quantity: *quantity,
        };
    }

    fn apply_ship(&self, ship: &mut Ship) {
        match self.action {
            Action::NORTH => ship.add_ship_y(self.quantity),
            Action::SOUTH => ship.add_ship_y(-1 * self.quantity),
            Action::EAST => ship.add_ship_x(self.quantity),
            Action::WEST => ship.add_ship_x(-1 * self.quantity),
            Action::LEFT => {
                let turns = self.quantity / 90;
                for turn in 0..turns {
                    ship.ship_left()
                }
            },
            Action::RIGHT => {
                let turns = self.quantity / 90;
                for turn in 0..turns {
                    ship.ship_right()
                }
            },
            Action::FORWARD => ship.add_ship_forward(self.quantity)
        }
    }

    fn apply_waypoint(&self, ship: &mut Ship) {
        match self.action {
            Action::NORTH => ship.add_waypoint_y(self.quantity),
            Action::SOUTH => ship.add_waypoint_y(-1 * self.quantity),
            Action::EAST => ship.add_waypoint_x(self.quantity),
            Action::WEST => ship.add_waypoint_x(-1 * self.quantity),
            Action::LEFT => {
                let turns = self.quantity / 90;
                for turn in 0..turns {
                    ship.waypoint_left()
                }
            },
            Action::RIGHT => {
                let turns = self.quantity / 90;
                for turn in 0..turns {
                    ship.waypoint_right()
                }
            },
            Action::FORWARD => ship.add_waypoint_forward(self.quantity)
        }
    }
}

fn puzzle1(lines: Vec<String>) -> u32 {
    let mut ship = Ship::new();

    lines.iter()
        .map(ShipInstruction::new)
        .for_each(|instruction| instruction.apply_ship(&mut ship));

    return ship.distance_from_origin();
}

fn puzzle2(lines: Vec<String>) -> u32 {

    let mut ship = Ship::new();

    lines.iter()
        .map(ShipInstruction::new)
        .for_each(|instruction| instruction.apply_waypoint(&mut ship));

    return ship.distance_from_origin();

}

#[cfg(test)]
mod tests {
    use crate::day12::puzzle1;
    use crate::day12::puzzle2;
    use crate::utils::read_lines;

    #[test]
    fn test_puzzle_1() {
        let strs = vec!["F10",
                        "N3",
                        "F7",
                        "R90",
                        "F11"];

        let strings = strs.into_iter()
            .map(String::from)
            .collect();

        let result = puzzle1(strings);
        assert_eq!(result, 25);
    }

    #[test]
    fn test_puzzle_1_2() {
        let strs = read_lines("data/Day12.txt").unwrap();

        let result = puzzle1(strs);
        assert_eq!(result, 381);
    }

    #[test]
    fn test_puzzle_2() {
        let strs = vec!["F10",
                        "N3",
                        "F7",
                        "R90",
                        "F11"];

        let strings = strs.into_iter()
            .map(String::from)
            .collect();

        let result = puzzle2(strings);
        assert_eq!(result, 286);
    }

    #[test]
    fn test_puzzle_2_2() {
        let strings = read_lines("data/Day12.txt").unwrap();

        let result = puzzle2(strings);
        assert_eq!(result, 28591);
    }
}